src/aspire/abinitio/commonline_d2.py

```diff
@@ -0,0 +1,1890 @@
+            )
+
+            # Compare to v_{ii}(r)=v_{ik}v_{ki}.
+            self_prods = Rijs_rows[ik] @ Rijs_rows_t[ik]
+            self_prods = self_prods.reshape(3, 9)
+
+            prod_arr1 = prod_arr.copy()
+            prod_arr1 -= self_prods[None, :, None]
+            norms1 = norms1 + np.sum(prod_arr1**2, axis=3)
+
+            prod_arr2 = prod_arr.copy()
+            prod_arr2 += self_prods[None, :, None]
+            norms2 = norms2 + np.sum(prod_arr2**2, axis=3)
+
+            norms = np.minimum(norms1, norms2)
+
+            for r in range(6):
+                p1 = trip_perms[r]
+                for s in range(6):
+                    p2 = trip_perms[s]
+                    m[r, s] = (
+                        norms[p2[0], p1[0], 0]
+                        + norms[p2[1], p1[1], 1]
+                        + norms[p2[2], p1[2], 2]
+                    )
+
+            # In the event of duplicate min values min_idx is the first occurence
+            # by column order to match matlab outputs.
+            min_idx = np.unravel_index(np.argmin(m.T), m.shape)[::-1]
+            votes[trip_idx] = m[min_idx]
+
+            # Store permutation indices as digits of a base 10 number.
+            colors_i[trip_idx, :2] = [
+                100 * (min_idx[0] + 1),
+                10 * (min_idx[1] + 1),
+            ]
+
+            # Calculate the relative permutation of Rik to Rij given
+            # by (sigma_ik)\circ(sigma_ij)^-1
+            inv_jk_perm = inverse_perms[min_idx[1]]
+            rel_perm = trip_perms[min_idx[0]]
+            rel_perm = rel_perm[inv_jk_perm]
+            colors_i[trip_idx, 2] = (2 * (rel_perm[0] + 1) - 1) + (
+                rel_perm[1] > rel_perm[2]
+            )
+            trip_idx += 1
+
+        colors_i = np.sum(colors_i, axis=1)
+
+        return colors_i
+
+    def _mult_cmat_by_vec(self, c_perms, v):
+        """
+        Multiply color matrix by vector v "on the fly".
+
+        :param c_perms: An (N over 3) vector. Each corresponds to a triplet of
+            indices i<j<k and indicates the relative permutation of tuples
+            {Qi^TIrQj},{Qj^TIlQk} and {Qk^TIlQi}. The color matrix can be
+            completely reconstructed from this information, which is used
+            here to execute a single multiplication of the matrix by the
+            vector v instead of explicitely computing and storing the prohibitively
+            large color matrix in memory.
+        :param v: vector to be multiplied by 'color matrix'.
+        """
+        t_perms = np.array(
+            [[0, 1, 2], [0, 2, 1], [1, 0, 2], [1, 2, 0], [2, 0, 1], [2, 1, 0]]
+        )
+        i_perms = np.array(
+            [[0, 1, 2], [0, 2, 1], [1, 0, 2], [2, 0, 1], [1, 2, 0], [2, 1, 0]]
+        )
+        out = np.zeros_like(v)
+        trip_idx = 0
+        for i in range(self.n_img):
+            for j in range(i + 1, self.n_img - 1):
+                ij = 3 * self.pairs_to_linear[i, j]
```

Comment:
3? Maybe rename or do the multiplication later.

Reply:
This is a vector of length 3 * (n_choose_2) where we update 3 entries for each pair ij. `3 * self.pairs_to_linear[i, j]` is just the first of the 3 entries. Struggling for a better name to use. How does `ij_block` sound? Not sure if there's analogous term to "block matrix" but for vectors.

I'll use that for now. Let me know if that's ok.

Reply:
Much better. Thanks!